ome/ome-files-cpp
# Add support for reading and writing OME-TIFF with sub-resolutions

Support reading OME-TIFF pyramids, as already implemented in Bio-Formats.

- Add two-level (series, resolution) `CoreMetadataList`
- Drop flat `CoreMetadataList` and associated `coreIndex` (unlike Bio-Formats, this doesn't have the same compatibility concerns, and it greatly simplifies the implementation)
- Split up the `initFile` method in `OMETIFFReader` into several helper functions; this makes it easier to reason about the scope and ordering of actions during setup, as well as allowing reuse of some of the helpers needed by the the sub-resolution setup
- Read and use the IFDs references by `SUBIFDS` to create the extra `CoreMetadata` resolution list elements
- `openBytes` will chose the IFD for the current series and resolution
- Update all unit tests to match the changed behaviour

Support writing OME-TIFF pyramids, similar to how it is implemented in https://github.com/openmicroscopy/bioformats/pull/3189

- Add helper functions to `MetadataTools` to add, query and remove resolution annotations on the metadata store
- Add sub-resolution methods to `FormatWriter`, and additionally expose some core metadata methods already implemented in `detail/FormatWriter`, so that e.g. SizeX and SizeY can be correctly introspected for the current resolution
- FormatWriter strips off the resolution annotations before writing is completed, so they don't get serialised
- TIFF wrapper changes include adding the sub-file type bit flags, adding subifd convenience methods to IFD, and adding support for writing IFDs to the TIFF interface, and flushing tiles in any order (needed for multi-threaded tile writing)
- OME-TIFF changes include adding IFD to the per-plane state (used by the reader), handling IFD and Sub-IFD flushing and switching when writing
- Unit test ensures that round-tripping of basic metadata and pixel data works.
- Example demonstrates rather more sophisticated use of pyramids, generating fractals at arbitrary scale, using multi-sampling for image quality and multi-threaded tile generation and writing to make it scale up to as many cores as you have free.

Other changes for safety, usability and performance:

- Classes hiding their internal state via `Impl` classes now use `unique_ptr` instead of `shared_ptr`.  This is both more efficient and safer since sharing of the impl object between instances is impossible.
- PixelBuffer size reduced from 9D to 4D; this has significant performance impact, speeding up every index calculation by 2.25×; iterating over pixel buffers is a significant bottleneck and this is a huge improvement.  It's also much easier and safer to use now, with there being two variants (interleaved, non-interleaved, rather than 12, where previously the six dimension orders were also taken into consideration).
- Unused helper functions dropped from OMETIFFWriter; removes some compiler warnings
- TileBuffer: Use `vector` rather than custom allocation of memory; it's safer and no less efficient

Sample files (uploaded to glencoe FTP server) for use as reference files and for testing:

- fractals-small.ome.tiff (baseline TIFF)
- fractals.ome.tiff (BigTIFF)

Testing:

- Run `ome-files info` on a pyramid OME-TIFF and check the resolution level count and dimensions match the expected number and sizes
- Run `ome-files view` with a pyramid OME-TIFF as documented in https://github.com/ome/ome-qtwidgets/pull/42 and check that the levels can be displayed correctly
- Run `./examples/subresolution fractal-example.ome.tiff` and check this with `tiffinfo [-o offset]`, `tiffgt [-o offset]`, `showinf -noflat -series n -resolution m` and `ome-files info` to check that the top-level IFDs and SubIFDs are present and correct and reported and viewed correctly by all the tools.  You might run into memory limits with `showinf` for the largest resolutions.
- Review the documentation addition on the tutorial page.

See also:

- https://github.com/ome/ome-qtwidgets/pull/42
- https://github.com/ome/ome-files-py/pull/30
- https://github.com/openmicroscopy/ome-files-performance/pull/62

/cc @dsudar
